From 1ea138f68f0b2388aa6f3b9d6e0d9208a15ccdd0 Mon Sep 17 00:00:00 2001
From: Eric Lundby <Eric.Lundby@gmail.com>
Date: Tue, 28 Oct 2025 09:20:13 -0600
Subject: [PATCH 1/2] 22

no such package '@@com_google_absl//absl/log': BUILD file not found in directory 'absl/log' of external repository @@com_google_absl.

no such target '@@com_google_absl//absl/base:nullability': target 'nullability' not declared in package 'absl/base'

symbol not found in flat namespace '__ZTVN4absl12lts_2025012714flags_internal8FlagImplE'

no such target '@@com_google_absl//absl/functional:overload': target 'overload' not declared in package 'absl/functional'

no such target '@@com_google_absl//absl/strings:string_view': target 'string_view' not declared in package 'absl/strings'

no such target '@@com_google_absl//absl/base:prefetch': target 'prefetch' not declared in package 'absl/base'

no such target '@@com_google_absl//absl/base:no_destructor': target 'no_destructor' not declared in package 'absl/base'

no such package '@@com_google_absl//absl/crc': BUILD file not found in directory 'absl/crc' of external repository @@com_google_absl

no such target '@@com_google_absl//absl/functional:overload': target 'overload' not declared in package 'absl/functional'

no such target '@@com_google_absl//absl/utility:if_constexpr': target 'if_constexpr' not declared in package 'absl/utility'

ld: library not found for -labsl_flags

_pywrap_tensorflow_internal.so, 0x000A): symbol not found in flat namespace '__ZTVN4absl12lts_2025012714flags_internal8FlagImplE'
---
 tensorflow/BUILD                              | 26 ++++++-
 tensorflow/python/BUILD                       |  8 +++
 .../third_party/absl/system.absl.base.BUILD   | 18 +++++
 .../third_party/absl/system.absl.crc.BUILD    | 33 +++++++++
 .../third_party/absl/system.absl.flags.BUILD  | 30 +++++---
 .../absl/system.absl.functional.BUILD         |  5 ++
 .../third_party/absl/system.absl.log.BUILD    | 71 +++++++++++++++++++
 .../absl/system.absl.strings.BUILD            | 14 +++-
 .../absl/system.absl.utility.BUILD            |  7 +-
 .../xla/third_party/absl/workspace.bzl        | 31 ++++++++
 10 files changed, 228 insertions(+), 15 deletions(-)
 create mode 100644 third_party/xla/third_party/absl/system.absl.crc.BUILD
 create mode 100644 third_party/xla/third_party/absl/system.absl.log.BUILD

diff --git a/tensorflow/BUILD b/tensorflow/BUILD
index 6e86e06902c..40af51839d1 100644
--- a/tensorflow/BUILD
+++ b/tensorflow/BUILD
@@ -1203,6 +1203,10 @@ tf_cc_shared_library(
             }),
     soversion = VERSION,
     static_deps = PACKAGE_STATIC_DEPS,
+    deps = [
+        "@com_google_absl//absl/flags:flag",
+        "@com_google_absl//absl/flags:reflection",
+    ],
     visibility = ["//visibility:public"],
 )
 
@@ -1292,8 +1296,14 @@ tf_cc_shared_library(
         "//tensorflow:tf_version_script.lds",
     ],
     dynamic_deps = select({
-        "//tensorflow:macos": ["//tensorflow:libtensorflow_framework.%s.dylib" % VERSION],
-        "//conditions:default": ["//tensorflow:libtensorflow_framework.so.%s" % VERSION],
+        "//tensorflow:macos": [
+            "//tensorflow:libtensorflow_framework.%s.dylib" % VERSION,
+            "@com_google_absl//absl/flags:flag",
+        ],
+        "//conditions:default": [
+            "//tensorflow:libtensorflow_framework.so.%s" % VERSION,
+            "@com_google_absl//absl/flags:flag",
+        ],
     }),
     exports_filter = [
         "//:__subpackages__",
@@ -1304,11 +1314,23 @@ tf_cc_shared_library(
     linkopts = select({
         "//tensorflow:macos": [
             "-Wl,-exported_symbols_list,$(location //tensorflow:tf_exported_symbols.lds)",
+            "-labsl_flags_internal",
+            "-labsl_flags_commandlineflag",
+            "-labsl_flags_commandlineflag_internal",
+            "-labsl_flags_config",
+            "-labsl_flags_marshalling",
+            "-labsl_flags_reflection",
         ],
         "//tensorflow:windows": [],
         "//conditions:default": [
             "-Wl,-z,defs",
             "-Wl,--version-script,$(location //tensorflow:tf_version_script.lds)",
+            "-labsl_flags_internal",
+            "-labsl_flags_commandlineflag",
+            "-labsl_flags_commandlineflag_internal",
+            "-labsl_flags_config",
+            "-labsl_flags_marshalling",
+            "-labsl_flags_reflection",
         ],
     }),
     per_os_targets = True,
diff --git a/tensorflow/python/BUILD b/tensorflow/python/BUILD
index 8fea090dc44..0a2a3761408 100644
--- a/tensorflow/python/BUILD
+++ b/tensorflow/python/BUILD
@@ -850,6 +850,14 @@ pywrap_tensorflow_macro(
         "@com_google_absl//absl/container:inlined_vector",
         "@com_google_absl//absl/types:optional",
         "@com_google_absl//absl/types:span",
+        "@com_google_absl//absl/flags:parse",
+        "@com_google_absl//absl/random/internal:pool_urbg",
+        "@com_google_absl//absl/flags:flag",
+        "@com_google_absl//absl/flags:reflection",
+        "@com_google_absl//absl/flags:config",
+        "@com_google_absl//absl/flags:marshalling",
+        "@com_google_absl//absl/flags:program_name",
+        "@com_google_absl//absl/flags:private_handle_accessor",
         "@local_xla//third_party/python_runtime:headers",
         "@local_xla//xla/backends/profiler/cpu:python_tracer",
         "@local_xla//xla/stream_executor:stream_executor_impl",
diff --git a/third_party/xla/third_party/absl/system.absl.base.BUILD b/third_party/xla/third_party/absl/system.absl.base.BUILD
index d6bf8748dee..9c9a2fa7852 100644
--- a/third_party/xla/third_party/absl/system.absl.base.BUILD
+++ b/third_party/xla/third_party/absl/system.absl.base.BUILD
@@ -13,6 +13,7 @@ package(default_visibility = ["//visibility:public"])
     "errno_saver",
     "fast_type_id",
     "pretty_function",
+    "no_destructor",
 ]]
 
 cc_library(
@@ -64,6 +65,23 @@ cc_library(
     ],
 )
 
+cc_library(
+    name = "nullability",
+    defines = [
+        "absl_nonnull=",
+        "absl_nullable=", 
+        "absl_nullability_unknown=",
+    ],
+)
+
+cc_library(
+    name = "prefetch",
+    deps = [
+        ":config",
+        ":core_headers",
+    ],
+)
+
 cc_library(
     name = "throw_delegate",
     linkopts = ["-labsl_throw_delegate"],
diff --git a/third_party/xla/third_party/absl/system.absl.crc.BUILD b/third_party/xla/third_party/absl/system.absl.crc.BUILD
new file mode 100644
index 00000000000..21ac1cf81ea
--- /dev/null
+++ b/third_party/xla/third_party/absl/system.absl.crc.BUILD
@@ -0,0 +1,33 @@
+load("@rules_cc//cc:defs.bzl", "cc_library")
+
+package(default_visibility = ["//visibility:public"])
+
+licenses(["notice"])  # Apache
+
+cc_library(
+    name = "crc",
+    linkopts = ["-labsl_crc"],
+    deps = [
+        "//absl/base:raw_logging_internal",
+    ],
+)
+
+cc_library(
+    name = "crc32c",
+    linkopts = ["-labsl_crc32c"],
+)
+
+cc_library(
+    name = "crc_cord_state",
+    linkopts = ["-labsl_crc_cord_state"],
+)
+
+cc_library(
+    name = "crc_internal",
+    linkopts = ["-labsl_crc_internal"],
+)
+
+cc_library(
+    name = "non_temporal_memcpy",
+    linkopts = ["-labsl_base"],
+)
\ No newline at end of file
diff --git a/third_party/xla/third_party/absl/system.absl.flags.BUILD b/third_party/xla/third_party/absl/system.absl.flags.BUILD
index aff653c7e5b..0f5d32a122d 100644
--- a/third_party/xla/third_party/absl/system.absl.flags.BUILD
+++ b/third_party/xla/third_party/absl/system.absl.flags.BUILD
@@ -5,9 +5,7 @@ package(default_visibility = ["//visibility:public"])
 cc_library(
     name = "program_name",
     linkopts = ["-labsl_flags_program_name"],
-    visibility = [
-        "//absl/flags:__pkg__",
-    ],
+    visibility = ["//visibility:public"],
     deps = [
         "//absl/strings",
         "//absl/synchronization",
@@ -52,9 +50,7 @@ cc_library(
 cc_library(
     name = "private_handle_accessor",
     linkopts = ["-labsl_flags_private_handle_accessor"],
-    visibility = [
-        "//absl/flags:__pkg__",
-    ],
+    visibility = ["//visibility:public"],
     deps = [
         ":commandlineflag",
         ":commandlineflag_internal",
@@ -79,7 +75,7 @@ cc_library(
 cc_library(
     name = "flag_internal",
     linkopts = ["-labsl_flags_internal"],
-    visibility = ["//absl/base:__subpackages__"],
+    visibility = ["//visibility:public"],
     deps = [
         ":commandlineflag",
         ":commandlineflag_internal",
@@ -97,11 +93,25 @@ cc_library(
 
 cc_library(
     name = "flag",
-    linkopts = ["-labsl_flags"],
+    linkopts = [
+        "-labsl_flags_internal",
+        "-labsl_flags_commandlineflag",
+        "-labsl_flags_commandlineflag_internal", 
+        "-labsl_flags_config",
+        "-labsl_flags_marshalling",
+        "-labsl_flags_reflection",
+        "-labsl_flags_private_handle_accessor",
+        "-labsl_flags_program_name",
+    ],
     deps = [
         ":config",
         ":flag_internal",
         ":reflection",
+        ":commandlineflag",
+        ":commandlineflag_internal",
+        ":marshalling",
+        ":private_handle_accessor",
+        ":program_name",
         "//absl/base",
         "//absl/strings",
     ],
@@ -110,9 +120,7 @@ cc_library(
 cc_library(
     name = "usage_internal",
     linkopts = ["-labsl_flags_usage_internal"],
-    visibility = [
-        "//absl/flags:__pkg__",
-    ],
+    visibility = ["//visibility:public"],
     deps = [
         ":commandlineflag",
         ":config",
diff --git a/third_party/xla/third_party/absl/system.absl.functional.BUILD b/third_party/xla/third_party/absl/system.absl.functional.BUILD
index 9439bd0ba22..a12a44668d1 100644
--- a/third_party/xla/third_party/absl/system.absl.functional.BUILD
+++ b/third_party/xla/third_party/absl/system.absl.functional.BUILD
@@ -13,3 +13,8 @@ cc_library(
 cc_library(
     name = "function_ref",
 )
+
+cc_library(
+    name = "overload",
+    linkopts = ["-labsl_base"],
+)
\ No newline at end of file
diff --git a/third_party/xla/third_party/absl/system.absl.log.BUILD b/third_party/xla/third_party/absl/system.absl.log.BUILD
new file mode 100644
index 00000000000..3cc52453752
--- /dev/null
+++ b/third_party/xla/third_party/absl/system.absl.log.BUILD
@@ -0,0 +1,71 @@
+load("@rules_cc//cc:defs.bzl", "cc_library")
+
+package(default_visibility = ["//visibility:public"])
+
+licenses(["notice"])  # Apache
+
+cc_library(
+    name = "absl_log",
+    linkopts = [
+        "-labsl_vlog_config_internal",
+        "-labsl_log_internal_conditions",
+        "-labsl_log_internal_check_op",
+        "-labsl_log_internal_message",
+        "-labsl_log_internal_nullguard",
+    ],
+)
+
+cc_library(
+    name = "absl_check",
+    linkopts = [
+        "-labsl_vlog_config_internal",
+        "-labsl_log_internal_check_op",
+        "-labsl_log_internal_message",
+        "-labsl_log_internal_nullguard",
+    ],
+)
+
+alias(
+    name = "log",
+    actual = ":absl_log",
+)
+
+alias(
+    name = "check",
+    actual = ":absl_check",
+)
+
+cc_library(
+    name = "die_if_null",
+    linkopts = ["-labsl_die_if_null"],
+)
+
+cc_library(
+    name = "log_entry",
+    linkopts = ["-labsl_log_internal_message"],
+)
+
+cc_library(
+    name = "log_sink",
+    linkopts = ["-labsl_log_sink"],
+)
+
+cc_library(
+    name = "log_sink_registry",
+    linkopts = ["-labsl_log_sink"],
+)
+
+cc_library(
+    name = "globals",
+    linkopts = ["-labsl_log_globals"],
+)
+
+cc_library(
+    name = "log_flags",
+    linkopts = ["-labsl_log_flags"],
+)
+
+cc_library(
+    name = "initialize",
+    linkopts = ["-labsl_log_initialize"],
+)
\ No newline at end of file
diff --git a/third_party/xla/third_party/absl/system.absl.strings.BUILD b/third_party/xla/third_party/absl/system.absl.strings.BUILD
index fa9a7a84f67..0b7ca228662 100644
--- a/third_party/xla/third_party/absl/system.absl.strings.BUILD
+++ b/third_party/xla/third_party/absl/system.absl.strings.BUILD
@@ -24,9 +24,21 @@ cc_library(
     ],
 )
 
+cc_library(
+    name = "string_view",
+    linkopts = ["-labsl_string_view"],
+    deps = [
+        "//absl/base",
+        "//absl/base:config",
+        "//absl/base:core_headers",
+        "//absl/base:nullability",
+        "//absl/base:throw_delegate",
+    ],
+)
+
 cc_library(
     name = "cord",
-    linkopts = ["-labsl_cord"],
+    linkopts = ["-labsl_cord", "-labsl_cordz_functions", "-labsl_cordz_info", "-labsl_cord_internal"],
     deps = [
         ":str_format",
         "//absl/container:compressed_tuple",
diff --git a/third_party/xla/third_party/absl/system.absl.utility.BUILD b/third_party/xla/third_party/absl/system.absl.utility.BUILD
index e15049e261c..e9c25fe3714 100644
--- a/third_party/xla/third_party/absl/system.absl.utility.BUILD
+++ b/third_party/xla/third_party/absl/system.absl.utility.BUILD
@@ -1,6 +1,11 @@
 load("@rules_cc//cc:defs.bzl", "cc_library")
 
+package(default_visibility = ["//visibility:public"])
+
 cc_library(
     name = "utility",
-    visibility = ["//visibility:public"],
 )
+
+cc_library(
+    name = "if_constexpr",
+)
\ No newline at end of file
diff --git a/third_party/xla/third_party/absl/workspace.bzl b/third_party/xla/third_party/absl/workspace.bzl
index 9144c5bd5b6..e24a2b08b9c 100644
--- a/third_party/xla/third_party/absl/workspace.bzl
+++ b/third_party/xla/third_party/absl/workspace.bzl
@@ -11,9 +11,40 @@ def repo():
     ABSL_SHA256 = "c397cd9cca3f71724a8ddf183e7fa71c19196eaafd1dc2a3c86d3a572613a807"
     # LINT.ThenChange(//tensorflow/lite/tools/cmake/modules/abseil-cpp.cmake)
 
+    SYS_DIRS = [
+        "algorithm",
+        "base",
+        "crc",
+        "log",
+        "cleanup",
+        "container",
+        "debugging",
+        "flags",
+        "functional",
+        "hash",
+        "memory",
+        "meta",
+        "numeric",
+        "random",
+        "status",
+        "strings",
+        "synchronization",
+        "time",
+        "types",
+        "utility",
+    ]
+
+    SYS_LINKS = {
+        "//third_party/absl:system.absl.{name}.BUILD".format(name = n): "absl/{name}/BUILD.bazel".format(name = n)
+        for n in SYS_DIRS
+    }
+
     tf_http_archive(
         name = "com_google_absl",
         sha256 = ABSL_SHA256,
+        build_file = "//third_party/absl:com_google_absl.BUILD",
+        system_build_file = "//third_party/absl:system.BUILD",
+        system_link_files = SYS_LINKS,
         strip_prefix = "abseil-cpp-{commit}".format(commit = ABSL_COMMIT),
         urls = tf_mirror_urls("https://github.com/abseil/abseil-cpp/archive/{commit}.tar.gz".format(commit = ABSL_COMMIT)),
         patch_file = [
-- 
2.50.1 (Apple Git-155)

